design/one-pager-package-image-config.md

```diff
@@ -0,0 +1,377 @@
+passing them individually through the installation APIs.
+
+### Prior Art on Image Verification
+
+Verification of image signatures is a common practice in the container ecosystem
+and doing so in a Kubernetes environment is not unique to Crossplane. The two
+outstanding solutions in this regard are [Policy Controller] and [Kyverno].
+Both projects provide a way to enforce policies at the admission level with the
+APIs they expose.
+
+Policy Controller is a project from [Sigstore], the same organization that
+maintains the [cosign] tool for signing and verifying container images. Through
+its `ClusterImagePolicy` API, Policy Controller allows users to define policies
+for verifying images with a wide range of options. As of today, Policy
+Controller only operates on Pods or the native Kubernetes resources resulting in
+a Pod, e.g. Deployments, StatefulSets, etc. There is no way to configure it to
+work with Crossplane packages.
+
+Kyverno, on the other hand, is a general-purpose policy engine for Kubernetes
+that allows users to define policies for any Kubernetes resource. Using its
+`ClusterPolicy` API, one can define [rules to verify images] and by leveraging
+the `imageExtractors` field, it is possible to configure it to work with
+Crossplane packages. See the [using Kyverno for package image verification]
+section for more details.
+
+Another example in this space is the project [Flux]. Compared to the previous
+APIs, this project provides a [minimalistic API to verify images] with the
+`OCIRepository` API, marking some parts of the API (e.g. keyless verification)
+as experimental.
+
+In Crossplane itself, we had [a previous proposal] aiming to solve signature
+verification for packages but mostly focusing on the implementation details,
+primarily for verification using a shared public key. The follow-up
+[implementation PR] extended the solution also to support keyless verification
+by mostly following the same solution / similar API as Flux by extending the
+Package installation APIs for verification settings, leaving the dependencies
+out of the scope. This proposal aims to provide a more comprehensive solution by
+introducing an API that can be extended to support various verification settings
+and also covering the ones coming as a dependency.
+
+## Goals
+
+- Allow configuration of credentials for package images and their dependencies
+  at runtime.
+- Enable users to set policies for verifying image signatures.
+- Define an image-centric configuration API that centralizes settings for
+  package images.
+
+## Proposal
+
+This document proposes a new API, `ImageConfig`, under the `pkg.crossplane.io`
+API group, that allows users to configure settings for package images. The API
+enables users to define rules for matching images and configuring how to
+interact with the registries hosting the images, including authentication and
+TLS settings. It also provides a way to define policies for verifying image
+signatures.
```

Comment:
Packages can actually have two images, the `xpkg` itself and the runtime one defined at `spec.image` in the meta definition contained in the first one. Is this meant to validate only the first one or both?
We could think of leaving the runtime one to other tools, given that it results in a deployment/pod which can be easily validated and enforced by other off-the-shelf tools without any major modification, but maybe we should handle that too.

However, if not defined, runtime packages default the runtime image to the `xpkg` one, therefore we still need to think about the fact that signatures are on a specific `sha`, so to avoid validating an image and the node pulling a different image, we should set the deployment's image to use the sha and not just the tag. The safest thing to do would be having a validating webhook for pods to ensure no one is setting a different image as the one we intended, but it might be considered out of scope for our threat model

Reply:
> Is this meant to validate only the first one or both?

So far, I was always considering validating the package image, which may or may not be the same as runtime image. I can double check what would it mean to validate runtime image at the same time if they differ, but, I agree that images of workloads could be validated separately with an additional tool, e.g. Kyverno or policy controller. This applies to the images of Crossplane itself. For example, flux has a documentation like this on how one can validate their own images: https://kyverno.io/policies/flux/verify-flux-images/verify-flux-images/

> I think pinning the sha at least at the Deployment level would be already something.

Good point. I'll see how it would look like during implementation.

Reply:
About the pinning, [kyverno](https://kyverno.io/docs/writing-policies/verify-images/sigstore/#verifying-image-signatures) does the same and so did other tools I met in the past, although they did that at the pod level: `The rule mutates matching images to add the [image digest](https://docs.docker.com/engine/reference/commandline/pull/#pull-an-image-by-digest-immutable-identifier), when mutateDigest is set to true (which is the default), if the digest is not already specified`